### ISSUE TITLE
Issues with `ttnn.max_pool2d` and `ttnn.conv2d` and their current workaround

### ISSUE THREAD

**Issue by jerrysky3** [tenstorrent/pytorch2.0_ttnn#385]
**Edited this issue to include a clearer issue and task list**

Here are some issues we run into with `ttnn.max_pool2d` and `ttnn.conv2d` when trying to use them in models:

- `ttnn.max_pool2d`
  - [ ] tenstorrent/tt-metal#15146 
  - [ ] tenstorrent/tt-metal#15147 
  - [x] https://github.com/tenstorrent/tt-metal/issues/14976
  - [x] https://github.com/tenstorrent/tt-metal/issues/13901
  - [ ] https://github.com/tenstorrent/tt-metal/issues/12099
  - [ ] OOM for some large input shapes (e.g. `(1, 64, 360, 640)`)
  - [ ] tenstorrent/tt-metal#15281 
- `ttnn.conv2d/ttnn.conv_transpose2d`
  - [x] tenstorrent/tt-metal#15145 
  - [x] tenstorrent/tt-metal#15280
  - [x] Crash with some large weight shapes ([#15032](https://github.com/tenstorrent/tt-metal/issues/15032))
  - [ ] OOM for some large input shapes
         - Weight is too large (e.g. 1024 * 1024 * 1 * 1) but the sharding number is very small (e.g. 2)
  - [x] tenstorrent/tt-metal#15282
  - [ ] https://github.com/tenstorrent/tt-metal/issues/15148

**Some backgrounds**
- Both `ttnn.max_pool2d` and `ttnn.conv2d` output tensors are sharded with padding and in the shape `(1, 1, NHW[padding], C)`. While it is possible to let the next op to directly use in the same layout, eventually we need to convert them back to the original pytorch tensor shape `(N, H, W, C)`. The permute and reshape here are tricky because of the sharding and padding
  - There is a workaround to reshape the output back to `(N, H, W, C)` but you can see in the examples below it is far from ideal. We need to move the tensor back to host to unpad and reshape
- The input and weight tensors for `ttnn.conv2d` are tricky to pass. The input tensor often have odd inner-most-dim so we need to use tile layout for now (as odd inner-most dim creates alignment problem in row-major layout without padding). But when I pass tiled-device input tensor to `ttnn.conv2d`, I often get "Divide by 0" error in `ttnn::operations::data_movement::detail::interleaved_to_sharded_multi_core`. Same for weight when it is a tiled device tensor but it triggers a different error (see the example below)
  - The root cause "Divide by 0" needs to debug. I suspect it is related to incorrect auto-sharding calculation
  - The root cause of tiled-device weight tensor error needs to debug
  - There is a workaround to pass input and weight tensors but it requires to move both of them back to host, which is not ideal

Currently there are some workarounds for all issues mentioned above so it is possible to lowering `ttnn.max_pool2d` and `ttnn.conv2d`, but the performance will not be too good

**`max_pool2d` example (shape is from `Autoencoder (conv)`):**

```py
import torch
import ttnn


def main(device):
    torch_tensor = torch.rand((1, 16, 28, 28), dtype=torch.bfloat16)
    expected_output = torch.max_pool2d(torch_tensor, (2, 2), (2, 2), (0, 0), (1, 1))

    input_tensor = ttnn.from_torch(torch_tensor, layout=ttnn.ROW_MAJOR_LAYOUT, device=device)
    input_tensor = ttnn.permute(input_tensor, (0, 2, 3, 1))
    input_tensor = ttnn.reshape(input_tensor, (1, 1, 28 * 28, 16))
    output_tensor = ttnn.max_pool2d(input_tensor, 1, 28, 28, 16, (2, 2), (2, 2), (0, 0), (1, 1))
    # Output shape: (1, 1, 196[244], 16), sharded
    output_tensor = ttnn.sharded_to_interleaved(output_tensor, memory_config=ttnn.L1_MEMORY_CONFIG)
    # Output shape: (1, 1, 196[244], 16), row-major
    # Ideally we want to reshape back directly into (1, 14, 14, 16), but it will crash currently. I suspect it is due to the padding 224
    # output_tensor = ttnn.reshape(output_tensor, (1, 14, 14, 16))

    # As a workaround, we can move tensor back to host to remove the padding and the following reshape works
    output_tensor = ttnn.from_device(output_tensor)
    # Unpad (1, 1, 196[244], 16) to (1, 1, 196, 16)
    output_tensor = output_tensor.unpad((0, 0, 0, 0), (1, 1, 196, 16))
    output_tensor = ttnn.reshape(output_tensor, (1, 14, 14, 16))
    output_tensor = ttnn.to_device(output_tensor, device=device)
    output_tensor = ttnn.permute(output_tensor, (0, 3, 1, 2))
    print(output_tensor.shape, expected_output.shape)

    assert torch.allclose(ttnn.to_torch(output_tensor), expected_output)


if __name__ == "__main__":
    device = ttnn.open_device(device_id=0, l1_small_size=65536)
    try:
        main(device)
    finally:
        ttnn.close_device(device)

```

**`conv2d` example (shape is taken from `Mnist`):**

```py
import torch
import ttnn


def main(device):
    torch_input = torch.rand((1, 1, 28, 28), dtype=torch.bfloat16)
    torch_weight = torch.rand((32, 1, 3, 3), dtype=torch.bfloat16)

    stride = 1
    padding = 0
    dilation = 1
    expected_output = torch.conv2d(torch_input, torch_weight, stride=stride, padding=padding, dilation=dilation, groups=1)

    input_tensor = ttnn.from_torch(torch_input, layout=ttnn.TILE_LAYOUT, device=device)
    input_tensor = ttnn.permute(input_tensor, (0, 2, 3, 1))
    # If input_tensor is on device in tile layout, it triggers "Divide by 0 error"
    input_tensor = ttnn.from_device(input_tensor)
    input_tensor = ttnn.to_layout(input_tensor, layout=ttnn.ROW_MAJOR_LAYOUT)

    weight_tensor = ttnn.from_torch(torch_weight, layout=ttnn.TILE_LAYOUT, device=device)
    # If weight_tensor is on device in tile layout, it triggers "Conv weight matrix shape is invalid"
    weight_tensor = ttnn.from_device(weight_tensor)
    weight_tensor = ttnn.to_layout(weight_tensor, layout=ttnn.ROW_MAJOR_LAYOUT)

    output_tensor, _, _, _, _ = ttnn.conv2d(
        input_tensor=input_tensor,
        weight_tensor=weight_tensor,
        batch_size=1,
        in_channels=1,
        out_channels=32,
        input_height=28,
        input_width=28,
        kernel_size=(3, 3),
        stride=(stride, stride),
        padding=(padding, padding),
        dilation=(dilation, dilation),
        device=device,
    )
    output_tensor = ttnn.sharded_to_interleaved(output_tensor, memory_config=ttnn.L1_MEMORY_CONFIG)
    # Output shape: (1, 1, 676[704], 32), tiled
    # Ideally we want to reshape back directly into (1, 28, 28, 32), but it will crash currently. I suspect it is due to the padding 704
    # output_tensor = ttnn.reshape(output_tensor, (1, 26, 26, 32))

     # As a workaround, we can move tensor back to host to remove the padding and the following reshape works
    output_tensor = ttnn.from_device(output_tensor)
    output_tensor = ttnn.to_layout(output_tensor, layout=ttnn.ROW_MAJOR_LAYOUT)
    # Unpad (1, 1, 676[704], 32) to (1, 1, 676, 32)
    output_tensor = output_tensor.unpad((0, 0, 0, 0), (1, 1, 676, 32))
    output_tensor = ttnn.reshape(output_tensor, (1, 26, 26, 32))
    output_tensor = ttnn.to_device(output_tensor, device=device)
    output_tensor = ttnn.permute(output_tensor, (0, 3, 1, 2))
    print(output_tensor.shape, expected_output.shape)

    # Currently there are some numerical issue, still debugging
    # assert torch.allclose(ttnn.to_torch(output_tensor), expected_output)


if __name__ == "__main__":
    device = ttnn.open_device(device_id=0, l1_small_size=65536)
    try:
        main(device)
    finally:
        ttnn.close_device(device)
```

**Comment by jerrysky3** [comment]
Got the PCC from conv2d: `0.9998183`, which looks fine to me

**Comment by jerrysky3** [comment]
Tested max_pool2d with the usages from all models, here are more issues found:

**Channel row size must be power of 2**

Some usages don't satisfy this constraint. For example:

```py
        # Row size (nchannels * bytes = 192) should be power of 2 (false).
        [
            "Tensor<[1, 96, 112, 112]> self = ?",
            "List[int] kernel_size = [3, 3]",
            "List[int] stride = [2, 2]",
            "List[int] padding = [1, 1]",
        ],
```

**Shard page size must currently have L1 aligned page size**

One usage doesn't satisfy this constraint:

```py
["Tensor<[1, 4, 14, 14]> self = ?", "List[int] kernel_size = [2, 2]", "List[int] stride = [2, 2]"]
```

**Out of L1 memory**

Some usages require too much L1 memory even after sharding. For example the shape below results in the error:

```
Out of Memory: Not enough space to allocate 39456000 B L1 buffer across 50 banks, where each bank needs to store 789120 B
```

```py
# OOM
        [
            "Tensor<[1, 64, 360, 640]> self = ?",
            "List[int] kernel_size = [3, 3]",
            "List[int] stride = [2, 2]",
            "List[int] padding = [1, 1]",
        ],
```

** Ceil mode is not supported**

torch max_pool2d has a `ceil_mode` option to allow sliding window to exceed the right and bottom bound, which is currently not supported in `ttnn.max_pool2d`.

**Comment by mywoodstock** [comment]
@jerrysky3 Some comments/responses:
 
* You can pass in `memory_config` parameter to Maxpool2D to let the output be saved as interleaved tensors (this simply calls the interleaved to sharded internally so you dont need to explicitly call it)
* The output of Maxpool2D has been kept as `[1, 1, NHW, C]` as thats the most direct output maxpool can calculate. Eventually you will get the tensor as `[N, H, W, C]` once we have the new tensor infra ready.
* We are currently revamping the weights pre-processing so that it is decoupled from `conv2d` call. This will let you more control over it, and `conv2d` will be able to use weights already on device.
* Maxpool requirement of channels to be pow of 2 is being relaxed currently and support for generic sizes should be available soon.
* Some limitations on the shard size will also be relaxed once we have the new tensor infra ready.
* Yes, `ceil_mode` is not yet supported. We will be adding the support soon.

**Comment by pavlejosipovic** [comment]
@ayerofieiev-tt This issue is not actionable for me, I suggest closing it?
